R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+    X <- log_likelihood(pars = pars, 
+                            data = data, 
+                            model_params = model_params,
+                            steps_per_day = steps_per_day, 
+                            pars_obs = pars_obs, 
+                            n_particles = n_particles
+    ) 
+    X
+  }
+  
+  
+  
+  # convert the current parameters into format easier for mcmc to deal with
+  curr_pars <- unlist(pars_init)
+  day_zero <- as.Date(data$date[1])
+  curr_pars[['start_date']] <- as.numeric(day_zero - pars_init$start_date) # convert to numeric
+  
+  
+  if(any(curr_pars < pars_min | curr_pars > pars_max)) {
+    stop('initial parameters are outside of specified range')
+  }
+  
+
+  ## calculate initial prior
+  curr_lprior <- calc_lprior(pars = curr_pars)
+  
+  # run particle filter on initial parameters
+  p_filter_est <- calc_ll(pars = curr_pars)
+  
+  
+  # checks on log_prior and log_likelihood functions
+  
+  if(length(curr_lprior) > 1) {
+    stop('log_prior must return a single numeric representing the log prior')
+  }
+  if(curr_lprior > 0 ) {
+    stop('log_prior must be negative or zero')
+  }
+  if(is.infinite(curr_lprior)) {
+    stop('initial parameters are not compatible with supplied prior')
+  }
+  
+  if(length(p_filter_est) != 2) {
+    stop('log_likelihood function must return a list containing elements log_likelihood and sample_state')
+  }
+  if(!setequal(names(p_filter_est), c('log_likelihood', 'sample_state'))) {
+    stop('log_likelihood function must return a list containing elements log_likelihood and sample_state')
+  }
+  if(length(p_filter_est$log_likelihood) > 1) {
+    stop('log_likelihood must be a single numeric representing the estimated log likelihood')
+  }
+  if(p_filter_est$log_likelihood > 0) {
+    stop('log_likelihood must be negative or zero')
+  }
+  if(any(p_filter_est$sample_state < 0)) {
+    stop('sample_state must be a vector of non-negative numbers')
+  }
+  
+  
+  # extract loglikelihhod estimate and sample state
+  # calculate posterior
+  curr_ll <- p_filter_est$log_likelihood
+  curr_lpost <- curr_lprior + curr_ll
+  
+  curr_ss <- p_filter_est$sample_state
+  
+
+  ## combine into a vector to add to the results at the end
+  out_init <- c(curr_pars, 
+                'log_prior' = curr_lprior, 
+                'log_likelihood' = curr_ll, 
+                'log_posterior' = curr_lpost) 
+  
+  ss_init <- curr_ss
+
+  
+  ## initialise output arrays
+  
+  res <- matrix(data = NA, 
+                nrow = n_mcmc, 
+                ncol = length(out_init), 
+                dimnames = list(NULL, names(out_init)))
+  
+  states <- matrix(data = NA,
+                   nrow = n_mcmc, 
+                   ncol = length(ss_init))
+  
+  
+  ## create function to reflect proposal boundaries at pars_min and pars_max
+  # this ensures the proposal is symetrical and we can simplify the MH step
+  
+  reflect_proposal <- function(x, floor, cap) {
```

Comment:
Make these lambdas/functionals into their own internal functions, as with `calc_loglikelihood`. The main function is quite long and this would make it easier to read

Reply:
I have taken the reflect_proposal and propose_parameters functions out of the pmcmc sampler and created internal functions. I intend to write tests for these functions so am leaving the conversation unresolved for now.

Reply:
added in tests